Failed to release 32bits connection ID

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

1. Start tidb-server
2. Connect to tidb-server
3. Disconnect to tidb-server

### 2. What did you expect to see? (Required)

Nothing is wrong.

### 3. What did you see instead (Required)

There are many error logs about release connection ID:

```
[2023/07/29 10:02:07.286 +08:00] [ERROR] [globalconn.go:311] ["failed to release 32bits connection ID"] [connectionID=2097154] [localConnID=1]
[2023/07/29 10:02:07.286 +08:00] [ERROR] [globalconn.go:311] ["failed to release 32bits connection ID"] [connectionID=2097154] [localConnID=1]
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

```
mysql> select tidb_version();
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                 |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v7.3.0-alpha-279-gf4f78a1db3-dirty
Edition: Community
Git Commit Hash: f4f78a1db3018f6026263ffa6190ac133c483da8
Git Branch: master
UTC Build Time: 2023-07-29 03:50:02
GoVersion: go1.20.4
Race Enabled: false
Check Table Before Drop: false
Store: unistore |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
```

--- a/server/conn.go
+++ b/server/conn.go
@@ -335,9 +335,14 @@ func (cc *clientConn) Close() error {
 	return closeConn(cc, connections)
 }
 
+// closeConn should be idempotent.
+// It will be called on the same `clientConn` more than once to avoid connection leak.
 func closeConn(cc *clientConn, connections int) error {
 	metrics.ConnGauge.Set(float64(connections))
-	cc.server.dom.ReleaseConnID(cc.connectionID)
+	if cc.connectionID > 0 {
+		cc.server.dom.ReleaseConnID(cc.connectionID)
+		cc.connectionID = 0
+	}
 	if cc.bufReadConn != nil {
 		err := cc.bufReadConn.Close()
 		if err != nil {

